Fix locale parameter check and add test

At1ass opened the pull request:
The gpinitsystem utility fails when incoming locale value doesn't
correspond character-by-character to the some line of locale -a output
[[1](https://github.com/greenplum-db/gpdb/blob/5cfd4791b3e327f733ba4a2832f627ef70f4aa91/gpMgmt/bin/gpinitsystem#L644-L653), [2](https://github.com/greenplum-db/gpdb/blob/5cfd4791b3e327f733ba4a2832f627ef70f4aa91/gpMgmt/bin/gpinitsystem#L656-L665), [3](https://github.com/greenplum-db/gpdb/blob/5cfd4791b3e327f733ba4a2832f627ef70f4aa91/gpMgmt/bin/lib/gp_bash_functions.sh#L179-L186)]

Because locale -a among other things [represents](https://github.com/bminor/glibc/blob/2d5ec6692f5746ccb11db60976a6481ef8e9d74f/locale/programs/locale.c#L440-L442) locales from the archive [with](https://github.com/bminor/glibc/blob/2d5ec6692f5746ccb11db60976a6481ef8e9d74f/locale/programs/locarchive.c#L1148-L1160 ) a normalized representation of encodings like en_US.UTF-8, then a locale with a
non-normalized encoding will not be found.

The current path adds locale normalization functionality and checks
to find matches between normalized locales and locale -a output

Commit:
Fix locale parameter check and add test

Add locale parameter correction in gpinitsystem.

seryozhasmirnov commented on gpMgmt/test/behave/mgmt_utils/gpinitsystem.feature:
```diff
@@ -17,6 +17,11 @@ Feature: gpinitsystem tests
         And gpconfig should print "Master  value: off" to stdout
         And gpconfig should print "Segment value: off" to stdout
 
+	Scenario: gpinitsystem exits with status 0 when the user set locale parameters 
```
It seems that spaces should be used for indentation instead of tabs.

maksm90 replied:
Trailing space too

Commit:
Add function for normalize codeset.

Add function for normalize all codesets in glibc form.

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -193,9 +193,21 @@ WARN_MARK="<<<<<"
+	else
+		echo $1
+	fi
+}
```
We have to post here the link to code fragment (or refer to function name of libc library) from which we export the logic of locale normalization.

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -193,9 +193,21 @@ WARN_MARK="<<<<<"
+NORMALIZE_CODESET (){
+	if [[ "$1" =~ .*".".* ]]
+	then
+		local normalize_locale=$(echo $1 | perl -pe "s/(\.[\w-]+@?)/\L\1/g; s/[^\w.@]//g; ")  
```
Trailing spaces

Commit:
Fix pull request issues

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -193,13 +193,30 @@ WARN_MARK="<<<<<"
+# https://sourceware.org/git/?p=glibc.git;a=blob;f=intl/l10nflist.c;h=078a450dfec21faf2d26dc5d0cb02158c1f23229;hb=HEAD
+#
+
+NORMALIZE_CODESET (){
```
- inject the word `locale` (as real argument for transformation) into function name (e.g. rename function to `NORMALIZE_CODESET_IN_LOCALE`)
- insert the space between `)` and `{`.

Commit:
Update normalize codeset function

Commit:
Remove trailing spaces

Remove trailing spaces in behave tests for gpinitsystem

Commit:
Update normalize codeset func

Fix issues in NORMALIZE_CODESET_IN_LOCALE func

Commit:
Revert "Remove trailing spaces"

This reverts commit 75ecbb5b999ea511aa63aecb945ebe90d8550636.

Commit:
Fix tests and normalize func

Fix tests, fix normalize func and fixissue with LC_ALL

Commit:
Fix tests

Add --ignore-warnings flag for success tests

Commit:
Fix PR issues

Fix PR issues and remove --ignore-warnings

maksm90 reviewed (dismissed):
@At1ass finish summing up commit message

maksm90 commented on gpMgmt/test/behave/mgmt_utils/gpinitsystem.feature:
```diff
@@ -17,6 +17,28 @@ Feature: gpinitsystem tests
+        When the user runs command "gpinitsystem -a -c ../gpAux/gpdemo/clusterConfigFile"
+        Then gpinitsystem should return a return code of 0
+        Given the user runs "gpstate"
+        Then gpstate should return a return code of 0
```
This test case is not relevant for 6X because `gpinitsystem` [sets](https://github.com/arenadata/gpdb/blob/ba91667ba5698367d057b26c392447f6db72fbd2/gpMgmt/bin/gpinitsystem#L303) default `en_EN.UTF-8` locale when another value was not specified in command line argument. Also it makes sense to fix help documentation entry for command line [option](https://github.com/arenadata/gpdb/blob/ba91667ba5698367d057b26c392447f6db72fbd2/gpMgmt/doc/gpinitsystem_help#L114-L119) `--locale` according to corresponding [xml doc entry](https://github.com/arenadata/gpdb/blob/ba91667ba5698367d057b26c392447f6db72fbd2/gpdb-doc/dita/utility_guide/ref/gpinitsystem.xml#L146-L158) to not confuse end users.

maksm90 replied:
@At1ass please, make required fix to help documentation of `gpinitsystem`

maksm90 commented on gpMgmt/test/behave/mgmt_utils/gpinitsystem.feature:
```diff
@@ -17,6 +17,28 @@ Feature: gpinitsystem tests
 
+    Scenario: gpinitsystem creates a cluster when the user set LC_ALL env variable
+        Given create demo cluster config
+        And the user runs command "export LC_ALL=en_US.UTF-8"
```
Instead of specifying `export` command use more appropriate [action](https://github.com/arenadata/gpdb/blob/ba91667ba5698367d057b26c392447f6db72fbd2/gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py#L1377-L1380).

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -193,13 +193,43 @@ WARN_MARK="<<<<<"
+	if [[ $len == 1 ]];
+	then
+		echo $1
+	else
```
The branch above might be redundant .

Commit:
Fix PR issues

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -193,13 +193,35 @@ WARN_MARK="<<<<<"
+NORMALIZE_CODESET_IN_LOCALE () {
+	local language_and_territory=$(echo $1 | perl -ne 'print for /(^.+?(?=\.|@|$))/s')
+	local codeset=$(echo $1 | perl -ne 'print for /((?<=\.).+?(?=@|$))/s')
+	local modifier=$(echo $1 | perl -ne 'print for /((?<=@).+(?=$))/s' )
```
The pattern `/((?<=@).+(?=$))/s` adds new line symbol (`0x0A`) to the end of matched substring

maksm90 replied:
As evidence:
```bash
echo 'en_US.UTF-8@123' | perl -ne 'print for /((?<=\.).+(?=@|$))/s' | hexdump -C
00000000  31 32 33 0a                                       |123.|
00000004
```
the new line is preserved from `echo` call, to remove it use `-n` option.

maksm90 replied:
Enclosing the end of line patter `$` into look-ahead assertion `(?=...)` looks redundant.

At1ass replied:
If we add ? for pattern - end of line will not be included

```
echo 'en_US.UTF-8@123' | perl -ne 'print for /((?<=@).+(?=$))/s' | hexdump -C
00000000  31 32 33 0a                                       |123.|
00000004

echo 'en_US.UTF-8@123' | perl -ne 'print for /((?<=@).+?(?=$))/s' | hexdump -C
00000000  31 32 33                                          |123|
00000003

```

At1ass replied:
But variant with add -n option for echo will look better

maksm90 replied:
> If we add ? for pattern - end of line will not be included

Yes, as another option. But IMO the lazy mode for pattern capture is redundant here.

Commit:
Fix script for modifier part

Fix script for modifier part and update
doc for gpinitsystem

maksm90 commented on gpMgmt/doc/gpinitsystem_help:
```diff
@@ -114,12 +114,10 @@ OPTIONS
+ locale set to default [en_US.UTF-8]. A locale identifier consists 
+ of a language identifier and a region identifier, and optionally a 
+ character set encoding. For example, sv_SE is Swedish as spoken in Sweden, 
+ en_US is U.S. English, and fr_CA is French Canadian. If more than one character 
```
All changed lines contain trailing space. Use [filestyle](https://github.com/aserebryakov/filestyle) plugin in vim to detect this early.

At1ass replied:
All changed lines and another lines contains whitespaces in original style

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -193,13 +193,35 @@ WARN_MARK="<<<<<"
 
+#
+# Simplified version of _nl_normalize_codeset from glibc
+#https://sourceware.org/git/?p=glibc.git;a=blob;f=intl/l10nflist.c;h=078a450dfec21faf2d26dc5d0cb02158c1f23229;hb=1305edd42c44fee6f8660734d2dfa4911ec755d6#l294
```
inject space after `#` to align comment line

maksm90 commented on gpMgmt/doc/gpinitsystem_help:
```diff
@@ -114,12 +114,10 @@ OPTIONS
- and a region identifier, and optionally a character set encoding. 
- For example, sv_SE is Swedish as spoken in Sweden, en_US is U.S. 
- English, and fr_CA is French Canadian. If more than one character 
+ locale set to default [en_US.UTF-8]. A locale identifier consists 
```
Square brackets are redundant here. I don't see in those file the such style of highlighting.

Commit:
Fix PR issues

Commit:
Update conditions for pass locale check

Add scenario for unnormalized locales in 'locale -a' output

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -193,13 +193,35 @@ WARN_MARK="<<<<<"
+		fi
+	done
+	return 0
 }
```
`IN_ARRAY()` looks as general function to find the first argument in the list of second one. Therefore, injection of caller-specific normalization routine in it seems ugly violating modularity.
As `IN_ARRAY()` is used only in finding of locale string in list of `locale -a` output I propose to rename it to smth like `FIND_LOCALE()` that takes just one locale argument or create new function with the same signature in `gpinitsystem` utility and move to there `NORMALIZE_CODESET_IN_LOCALE()` function.

Commit:
Update IN_ARRAY func

Rename IN_ARRAY to LOCALE_IS_AVAILABLE.
LOCALE_IS_AVAILABLE now func with one parameter - locale for check
'locale -a' call now inside LOCALE_IS_AVAILABLE func

maksm90 commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -213,10 +213,11 @@ NORMALIZE_CODESET_IN_LOCALE () {
 
-IN_ARRAY () {
+LOCALE_IS_AVAILABLE () {
 	local locale=$(NORMALIZE_CODESET_IN_LOCALE $1)
```
Rename `locale` variable to smth more appropriate form so it's meaning doesn't conflict with `$1`

Commit:
Rename locale variable
